File: demos_code_J2_conception_objet/02_injection_constructor_vs_locator/DemoInjection.java
import java.util.HashMap;
import java.util.Map;

public class DemoInjection {
    interface Clock { long now(); }

    static final class SystemClock implements Clock {
        public long now() { return System.currentTimeMillis(); }
    }

    static final class ServiceLocator {
        private static final Map<Class<?>, Object> services = new HashMap<>();
        static <T> void register(Class<T> type, T instance) { services.put(type, instance); }
        static <T> T get(Class<T> type) { return type.cast(services.get(type)); }
    }

    static final class TokenServiceWithLocator {
        String createToken(String user) {
            // Dépendance invisible dans le constructeur et l'API de la classe.
            long timestamp = ServiceLocator.get(Clock.class).now();
            return user + "-" + timestamp;
        }
    }

    static final class TokenServiceWithInjection {
        private final Clock clock;
        TokenServiceWithInjection(Clock clock) { this.clock = clock; }
        String createToken(String user) { return user + "-" + clock.now(); }
    }

    public static void main(String[] args) {
        System.out.println("=== Service Locator ===");
        ServiceLocator.register(Clock.class, new SystemClock());
        System.out.println(new TokenServiceWithLocator().createToken("alice"));

        System.out.println("\n=== Injection constructeur ===");
        Clock fixedClock = () -> 1_700_000_000_000L;
        TokenServiceWithInjection service = new TokenServiceWithInjection(fixedClock);
        System.out.println(service.createToken("alice"));
        System.out.println("Même entrée, même sortie : " + service.createToken("alice"));
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/02_injection_constructor_vs_locator/demo_injection.py
from abc import ABC, abstractmethod
import time
from typing import Any, TypeVar, Type


class Clock(ABC):
    @abstractmethod
    def now(self) -> int:
        raise NotImplementedError


class SystemClock(Clock):
    def now(self) -> int:
        return int(time.time() * 1000)


class FixedClock(Clock):
    def __init__(self, value: int) -> None:
        self._value = value

    def now(self) -> int:
        return self._value


T = TypeVar("T")


class ServiceLocator:
    _services: dict[type, object] = {}

    @classmethod
    def register(cls, service_type: Type[T], instance: T) -> None:
        cls._services[service_type] = instance

    @classmethod
    def get(cls, service_type: Type[T]) -> T:
        service = cls._services.get(service_type)
        if service is None:
            raise RuntimeError(f"Service non enregistré : {service_type.__name__}")
        return service  # type: ignore[return-value]


class TokenServiceWithLocator:
    def create_token(self, user: str) -> str:
        # Dépendance invisible dans le constructeur et l'API de la classe.
        timestamp = ServiceLocator.get(Clock).now()
        return f"{user}-{timestamp}"


class TokenServiceWithInjection:
    def __init__(self, clock: Clock) -> None:
        self._clock = clock

    def create_token(self, user: str) -> str:
        return f"{user}-{self._clock.now()}"


if __name__ == "__main__":
    print("=== Service Locator ===")
    ServiceLocator.register(Clock, SystemClock())
    print(TokenServiceWithLocator().create_token("alice"))

    print("\n=== Injection constructeur ===")
    fixed_clock = FixedClock(1_700_000_000_000)
    service = TokenServiceWithInjection(fixed_clock)
    print(service.create_token("alice"))
    print("Même entrée, même sortie : " + service.create_token("alice"))
